core/src/cluster_nodes.rs

```diff
@@ -0,0 +1,441 @@
+            .position(|i| self.nodes[*i].pubkey() == self.pubkey)
+            .unwrap();
+        let (neighbors, children) = compute_retransmit_peers(fanout, self_index, &index);
+        // Assert that the node itself is included in the set of neighbors, at
+        // the right offset.
+        debug_assert_eq!(
+            self.nodes[neighbors[self_index % fanout]].pubkey(),
+            self.pubkey
+        );
+        let get_contact_infos = |index: Vec<usize>| -> Vec<&ContactInfo> {
+            index
+                .into_iter()
+                .map(|i| self.nodes[i].contact_info().unwrap())
+                .collect()
+        };
+        (get_contact_infos(neighbors), get_contact_infos(children))
+    }
+}
+
+fn new_cluster_nodes<T: 'static>(
+    cluster_info: &ClusterInfo,
+    stakes: &HashMap<Pubkey, u64>,
+) -> ClusterNodes<T> {
+    let self_pubkey = cluster_info.id();
+    let nodes = get_nodes(cluster_info, stakes);
+    let broadcast = TypeId::of::<T>() == TypeId::of::<BroadcastStage>();
+    // For backward compatibility:
+    //   * nodes which do not have contact-info are excluded.
+    //   * stakes are floored at 1.
+    // The sorting key here should be equivalent to
+    // ClusterInfo::sorted_stakes_with_index.
+    // Leader itself is excluded when sampling broadcast peers.
+    let index = nodes
+        .iter()
+        .enumerate()
+        .filter(|(_, node)| node.contact_info().is_some())
+        .filter(|(_, node)| !broadcast || node.pubkey() != self_pubkey)
+        .sorted_by_key(|(_, node)| Reverse((node.stake.max(1), node.pubkey())))
+        .map(|(index, node)| (node.stake.max(1), index))
+        .collect();
+    ClusterNodes {
+        pubkey: self_pubkey,
+        nodes,
+        index,
+        _phantom: PhantomData::default(),
+    }
+}
+
+// All staked nodes + other known tvu-peers + the node itself;
+// sorted by (stake, pubkey) in descending order.
+fn get_nodes(cluster_info: &ClusterInfo, stakes: &HashMap<Pubkey, u64>) -> Vec<Node> {
+    let self_pubkey = cluster_info.id();
+    // The local node itself.
+    std::iter::once({
+        let stake = stakes.get(&self_pubkey).copied().unwrap_or_default();
+        let node = NodeId::from(cluster_info.my_contact_info());
+        Node { node, stake }
+    })
+    // All known tvu-peers from gossip.
+    .chain(cluster_info.tvu_peers().into_iter().map(|node| {
+        let stake = stakes.get(&node.id).copied().unwrap_or_default();
+        let node = NodeId::from(node);
+        Node { node, stake }
+    }))
+    // All staked nodes.
+    .chain(
+        stakes
+            .iter()
+            .filter(|(_, stake)| **stake > 0)
+            .map(|(&pubkey, &stake)| Node {
+                node: NodeId::from(pubkey),
+                stake,
+            }),
+    )
+    .sorted_by_key(|node| Reverse((node.stake, node.pubkey())))
+    // Since sorted_by_key is stable, in case of duplicates, this
+    // will keep nodes with contact-info.
```

Comment:
This is because in the enum, the `ContactInfo` variant is declared before the `Pubkey` variant?

Reply:
No, it is because in the above iterator `chain`, all `NodeId:ContactInfo` come before all `NodeId:Pubkey`

Reply:
Ah yup, make sense, thanks!